Common/Algorithm/Framework/Alphas/Analysis/ReadOnlySecurityValuesCollection.cs

```diff
@@ -34,13 +36,49 @@ public class ReadOnlySecurityValuesCollection
         public ReadOnlySecurityValuesCollection(Dictionary<Symbol, SecurityValues> securityValuesBySymbol)
         {
             _securityValuesBySymbol = securityValuesBySymbol;
+            _securityValuesBySymbolFunc = null;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ReadOnlySecurityValuesCollection"/> class
+        /// </summary>
+        /// <remarks>This constructor has performance in mind. Only create the <see cref="SecurityValues"/>
+        /// for a <see cref="Symbol"/> if requested by a consumer.</remarks>
+        /// <param name="securityValuesBySymbolFunc">Function used to get the
+        /// <see cref="SecurityValues"/> for a specified <see cref="Symbol"/></param>
+        public ReadOnlySecurityValuesCollection(Func<Symbol, SecurityValues> securityValuesBySymbolFunc)
+        {
+            _securityValuesBySymbolFunc = securityValuesBySymbolFunc;
+            // lets be lazy for constructing the dictionary too!
+            _securityValuesBySymbol = null;
```

Comment:
Looks like we're overloading this type, perhaps an interface and a new type should be defined? In one case we have a read-only collection and in another case we have a lazy initialized collection. Might not be worth it, but having two responsibilities in this class does make it more complicated.

Reply:
Yeah, good point thought about it, but the class isn't really doing anything, pretty small, so I decided to make him a bit more complicated 😄